Implement applyProjection for JDBC connectors

martint opened the pull request.

martint committed:
Allow table metadata with no columns

Since the introduction of the applyXXX methods, a TableScan
represents a subquery, not just a raw table. It is possible that
such a subquery might have no columns due to a projection being
applied by the optimizer.

martint committed:
Implement applyProjection for JDBC connectors

electrum commented:
I was curious how skipping the check for no columns works and found that `QueryBuilder.getProjection()` does a `SELECT null` for that case.

electrum commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java:
```diff
@@ -107,12 +111,45 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
+                        handle.getLimit()),
+                projections,
+                assignments.entrySet().stream()
+                        .map(assignment -> new ProjectionApplicationResult.Assignment(assignment.getKey(), assignment.getValue(), ((JdbcColumnHandle) assignment.getValue()).getColumnType()))
```
Wrapping the `Assignment` arguments would make this easier to read. Maybe also static import it.

electrum commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcTableHandle.java:
```diff
@@ -125,6 +137,7 @@ public String toString()
         StringBuilder builder = new StringBuilder();
         builder.append(schemaTableName).append(" ");
         Joiner.on(".").skipNulls().appendTo(builder, catalogName, schemaName, tableName);
+        builder.append(" columns=").append(columns);
```
This will print like
```
columns=Optional.empty
```
or
```
columns=Optional[[orderkey:bigint:bigint],...]
```
It would look nicer to skip when not present
```java
columns.ifPresent(value -> builder.append(" columns=").append(value));

findepi commented on presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:
```diff
@@ -480,9 +480,6 @@ public TableMetadata getTableMetadata(Session session, TableHandle tableHandle)
         ConnectorMetadata metadata = getMetadata(session, catalogName);
         ConnectorTableMetadata tableMetadata = metadata.getTableMetadata(session.toConnectorSession(catalogName), tableHandle.getConnectorHandle());
-        if (tableMetadata.getColumns().isEmpty()) {
-            throw new PrestoException(NOT_SUPPORTED, "Table has no columns: " + tableHandle);
```
How will `SELECT *` work for such a table?
Do we have a test for this?

(IIRC i added the check here to prevent obscure errors later on, but this may be outdated.)

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java:
```diff
@@ -107,12 +112,48 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
+
+        List<JdbcColumnHandle> newColumns = assignments.values().stream()
+                .map(JdbcColumnHandle.class::cast)
+                .collect(Collectors.toList());
```
ImmutableList

kokosing replied:
`io.prestosql.util.MoreLists#mappedCopy` (requires move to plugin toolkit)

martint replied:
> io.prestosql.util.MoreLists#mappedCopy (requires move to plugin toolkit)

That's overkill for such a simple usage.

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcTableHandle.java:
```diff
@@ -51,13 +55,15 @@ public JdbcTableHandle(
         this.catalogName = catalogName;
         this.schemaName = schemaName;
         this.tableName = requireNonNull(tableName, "tableName is null");
+        this.columns = columns.map(ImmutableList::copyOf);
```
rnn `columns`

martint replied:
It’s unnecessary. The call to copy the list will fail eagerly with an NPE anyway. There’s no danger of holding on to an invalid value that will later cause the program to fail - this is the actual goal of validating arguments eagerly.

findepi replied:
yes, but it would fail with NPE without clear message, so we still prefer rnn

martint replied:
You look at the stacktrace see it (since you're going to be debugging anyway). The original motivation for adding these when we started using them in the early years of airlift was what I described above, but ok. One day when we move to Java 14+ we can get rid of all these usages.

findepi reviewed (commented):
How can we test that?

martint commented:
How can we test what, specifically?

kokosing reviewed (commented):
No tests.

You are still using old columns in `io.prestosql.plugin.jdbc.JdbcRecordSet#JdbcRecordSet`, so it is hard to tell if it working at all. If you would use columns from table handle in `JdbcRecordSet` then existing JDBC connectors would need to rely on `applyProjection` as well as other JDBC connectors that do not use `JdbcRecordSet`. Also then your change would be tested with existing tests.

kokosing reviewed (changes requested):
See my previous comment % I can approve this pull request as work in progress.

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcTableHandle.java:
```diff
@@ -88,6 +94,12 @@ public String getTableName()
     }
 
+    @JsonProperty
+    public Optional<List<JdbcColumnHandle>> getColumns()
```
add verification in `io.prestosql.plugin.jdbc.BaseJdbcClient#buildSql` that table handle's columns are same as the one passed explicitly

martint replied:
I added it to JdbcRecordsetProvider

martint commented:
> No tests.

Unfortunately, there's no infrastructure to test this explicitly, and there are no connectors that take advantage of this information during split generation.

> You are still using old columns in io.prestosql.plugin.jdbc.JdbcRecordSet#JdbcRecordSet, so it is hard to tell if it working at all. If you would use columns from table handle in JdbcRecordSet then existing JDBC connectors would need to rely on applyProjection as well as other JDBC connectors that do not use JdbcRecordSet. Also then your change would be tested with existing tests.

I added a check to verify that if the columns are present, they must match. I also added a comment explaining why it works the way it does currently.

kokosing commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcRecordSetProvider.java:
```diff
@@ -45,6 +47,13 @@ public RecordSet getRecordSet(ConnectorTransactionHandle transaction, ConnectorS
+        // If no columns are recorded, it means that applyProjection never got called (e.g., in the case all columns are being used) and all
+        // table columns should be returned. TODO: this is something that should be addressed once the getRecordSet API is revamped
+        jdbcTable.getColumns()
+                .ifPresent(tableColumns -> verify(ImmutableSet.copyOf(columns).equals(ImmutableSet.copyOf(tableColumns))));
```
what about the order?

martint replied:
Yes, good catch. I didn't recall if we made any guarantees, but the LocalExecutionPlanner respects the order of the output symbols of the TableScan. I'll update to check that, too.

kokosing reviewed (approved):
% comment

kokosing commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java:
```diff
@@ -322,4 +363,9 @@ public void dropSchema(ConnectorSession session, String schemaName)
+
+    private static boolean containSameElements(Iterable<? extends ColumnHandle> first, Iterable<? extends ColumnHandle> second)
+    {
+        return ImmutableSet.copyOf(first).equals(ImmutableSet.copyOf(second));
```
What about order here?

martint replied:
I can change it, but it doesn't really matter here. There's not much benefit in accepting a pushdown if it only changes the order of the fields (other than eliminating an almost-free projection from the plan)

martint replied:
Actually, we have to keep it like this for now. There's something going on in PushProjectionsIntoTableScan that's not preserving the order as it extracts the expressions to project. As a result, the plan never converges. I'll look into it separately and change this later when it's fixed.
